alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/navigationstrategies/KnownDestinationReaching.kt

```diff
@@ -0,0 +1,140 @@
+ * map and consists of a list of landmarks, it is computed so as to minimize the distance between the first
+ * landmark and the pedestrian's position and the distance between the last landmark and the destination. In
+ * spite of this, the first landmark may be far from the pedestrian's position and the same applies to the
+ * last landmark and the destination. This means the pedestrian may not have information regarding any path
+ * leading from its current position to the first landmark and from the last one to the destination (such a
+ * path may not exist at all).
+ * If the cognitive map is poor the returned path may be poor. On average when the first landmark is farther
+ * from the pedestrian's position than the destination itself it's just more convenient to pursue the latter.
+ * If the cognitive map is empty the path will be empty as well.
+ */
+fun <T, P, N : ConvexGeometricShape<P, *>, M : ConvexGeometricShape<P, *>> findKnownPath(
+    destination: P,
+    pedestrian: OrientingPedestrian<T, P, *, N, *>,
+    environment: EnvironmentWithGraph<*, T, P, *, M, *>
+): List<N> where P : Position<P>, P : Vector<P> = pedestrian.cognitiveMap.let { cognitiveMap ->
+    emptyList<N>().takeIf { cognitiveMap.vertexSet().isEmpty() } ?: run {
+        val currPos = environment.getPosition(pedestrian)
+        val currRoom = environment.graph.nodeContaining(currPos)
+        val destRoom = environment.graph.nodeContaining(destination)
+        if (currRoom == null || destRoom == null) {
+            return emptyList()
+        }
+        val landmarksIn: (room: M) -> Sequence<N> = { room ->
+            cognitiveMap.vertexSet().asSequence().filter { room.contains(it.centroid) }
+        }
+        val landmarksInAny: (rooms: List<M>) -> Sequence<N> = { rooms -> rooms.asSequence().flatMap(landmarksIn) }
+        val landmarksInCurrRoom = landmarksIn(currRoom)
+        val landmarksInDestRoom = landmarksIn(destRoom)
+        val landmarksAdjacentToCurrRoom = landmarksInAny(Graphs.successorListOf(environment.graph, currRoom))
+        val landmarksAdjacentToDestRoom = landmarksInAny(Graphs.predecessorListOf(environment.graph, destRoom))
+        /*
+         * In order, we try to find a path in which:
+         * - the first landmark is in current room and the last one in destination room
+         * - the first landmark is in current room and the last one in a room adjacent to destination room
+         * - the first landmark is in a room adjacent to current room and the last one in destination room
+         * - the first landmark is in a room adjacent to current room and the last one in a room adjacent to destination
+         * room
+         * - landmarks may be located anywhere (and may coincide), but we try to minimise the distance between the first
+         * landmark and currPos and the distance between the last landmark and destination
+         */
+        return (cartesianProduct(landmarksInCurrRoom, landmarksInDestRoom) +
```

Comment:
can you factor out this complicate return into its own function? Possibly two functions, one for building the sequence, and one for processing it

Reply:
Created two functions (more or less) as indicated.